Add meta function for native_multi_head_attention

haowu14 opened the pull request:
Test Plan:
lowering passes
```
TORCH_LOGS="+dynamo" TORCHDYNAMO_VERBOSE=1 RECURSIVE_PM=1 buck2 run @//mode/opt @//mode/inplace -c fbcode.enable_gpu_sections=true -c fbcode.platform=platform010 -c fbcode.nvcc_arch=a100,h100 -c fbcode.split-dwarf=true -c fbcode.dwp=true -c fbcode.enable_distributed_thinlto=true -c fbcode.use_link_groups=true inference_enablement/model_processing/infra/components/lowering/re:re_cinder -- -r '{"aot_inductor": {"serialized_inference_model_input_path":"ads_storage_fblearner/tree/user/facebook/fblearner/predictor/771029565/0/gpu_lowering/input.predictor.precompute.mix.gpu.merge","serialized_inference_model_output_path":"ads_storage_fblearner/tree/user/facebook/fblearner/predictor/771029565/0/gpu_lowering/sm90_arm_output.predictor.precompute.mix.gpu.merge","submodule_names_to_lower":["merge"],"inductor_lowering_context":{"aot_inductor_lowering_settings":{"max_batch_size":8192,"min_acc_module_size":10,"workdir":"/tmp/inductor_lowering_i_9kda5t","name":"gmpp_gpu_inductor_lowering","dll_name":"inductor_engine.so","use_scripting":true,"preset_lowerer":"","precision":1,"output_precision":1,"remote_cache_file_path_folder":"ads_storage_fblearner/tree/user/facebook/fblearner/predictor/771029565/","save_remote_cache":true,"aot_inductor_config":"{'\''max_autotune'\'': True, '\''comprehensive_padding'\'': False}","disable_dynamic_shapes":false,"remove_unexpected_type_cast":true,"disable_constraint_solver":false,"sample_input_tile_factor":1,"disable_acc_tracer":false,"generate_sample_inputs":true,"tile_sample_input_by_dynamic_shape":false,"node_replacement_dict":"","auto_dynamic_shapes":false,"auto_dynamic_shapes_min_size":1,"auto_dynamic_shapes_max_size":1048576,"max_acc_splits":-1,"dynamic_size":-1,"pre_dispatch_export":true,"merge_split_optimization":false}},"model_entity_id":771029565,"model_snapshot_id":0,"add_sample_inputs":true,"hardware_type":0,"platform_arch":2,"lowering_lib_pkg":"ien.lower.cg1:132"}}'
```
E2E test: f787959188

Rollback Plan:

Reviewed By: henryoier

Differential Revision: D80179945

ezyang reviewed (approved):
Looks reasonable. I was hoping an OpInfo test for this would start passing. @drisspg do we have one? Please also double check that the strides of the output are in fact contiguous.

ezyang commented:
un-approving as CI is failing, rightfully so!

haowu14 committed:
Add meta function for native_multi_head_attention (#161991)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/161991

Test Plan:
lowering passes
```
TORCH_LOGS="+dynamo" TORCHDYNAMO_VERBOSE=1 RECURSIVE_PM=1 buck2 run @//mode/opt @//mode/inplace -c fbcode.enable_gpu_sections=true -c fbcode.platform=platform010 -c fbcode.nvcc_arch=a100,h100 -c fbcode.split-dwarf=true -c fbcode.dwp=true -c fbcode.enable_distributed_thinlto=true -c fbcode.use_link_groups=true inference_enablement/model_processing/infra/components/lowering/re:re_cinder -- -r '{"aot_inductor": {"serialized_inference_model_input_path":"ads_storage_fblearner/tree/user/facebook/fblearner/predictor/771029565/0/gpu_lowering/input.predictor.precompute.mix.gpu.merge","serialized_inference_model_output_path":"ads_storage_fblearner/tree/user/facebook/fblearner/predictor/771029565/0/gpu_lowering/sm90_arm_output.predictor.precompute.mix.gpu.merge","submodule_names_to_lower":["merge"],"inductor_lowering_context":{"aot_inductor_lowering_settings":{"max_batch_size":8192,"min_acc_module_size":10,"workdir":"/tmp/inductor_lowering_i_9kda5t","name":"gmpp_gpu_inductor_lowering","dll_name":"inductor_engine.so","use_scripting":true,"preset_lowerer":"","precision":1,"output_precision":1,"remote_cache_file_path_folder":"ads_storage_fblearner/tree/user/facebook/fblearner/predictor/771029565/","save_remote_cache":true,"aot_inductor_config":"{'\''max_autotune'\'': True, '\''comprehensive_padding'\'': False}","disable_dynamic_shapes":false,"remove_unexpected_type_cast":true,"disable_constraint_solver":false,"sample_input_tile_factor":1,"disable_acc_tracer":false,"generate_sample_inputs":true,"tile_sample_input_by_dynamic_shape":false,"node_replacement_dict":"","auto_dynamic_shapes":false,"auto_dynamic_shapes_min_size":1,"auto_dynamic_shapes_max_size":1048576,"max_acc_splits":-1,"dynamic_size":-1,"pre_dispatch_export":true,"merge_split_optimization":false}},"model_entity_id":771029565,"model_snapshot_id":0,"add_sample_inputs":true,"hardware_type":0,"platform_arch":2,"lowering_lib_pkg":"ien.lower.cg1:132"}}'
```
E2E test: f787959188

Reviewed By: henryoier, ezyang

Differential Revision: D80179945
